src/Microsoft.AspNetCore.WebUtilities/Base64UrlTextEncoder.cs

```diff
@@ -26,17 +29,76 @@ public static string Encode(byte[] data)
         /// <returns>The decoded data.</returns>
         public static byte[] Decode(string text)
         {
-            return Convert.FromBase64String(Pad(text.Replace('-', '+').Replace('_', '/')));
+            return Convert.FromBase64String(DecodeToBase64String(text));
         }
 
-        private static string Pad(string text)
+        // To enable unit testing
+        internal static string EncodeInternal(string base64EncodedString)
         {
-            var padding = 3 - ((text.Length + 3) % 4);
-            if (padding == 0)
+            var length = base64EncodedString.Length;
+            while (length > 0 && base64EncodedString[length - 1] == '=')
+            {
+                length--;
+            }
+
+            if (length == 0)
+            {
+                return string.Empty;
+            }
+
+            var inplaceStringBuilder = new InplaceStringBuilder(length);
```

Comment:
The issue there is that you end up allocating two things (the char[] and the string). `InplaceStringBuilder` modifies the backing buffer for a string so you avoid the first allocation.